src/java.base/share/classes/sun/security/ssl/SSLSessionImpl.java

```diff
@@ -1456,6 +1562,259 @@ public List<SNIServerName> getRequestedServerNames() {
+
+            // Unlikely, but check anyway.
+            if (exporterMasterSecret == null) {
+                throw new RuntimeException(
+                        "Exporter master secret not captured");
+            }
+
+            // Do RFC 8446:7.1-7.5 calculations
+
+            /*
+             * TLS-Exporter(label, context_value, key_length) =
+             *     HKDF-Expand-Label(Derive-Secret(Secret, label, ""),
+             *         "exporter", Hash(context_value), key_length)
+             *
+             * Derive-Secret(Secret, Label, Messages) =
+             *     HKDF-Expand-Label(Secret, Label,
+             *         Transcript-Hash(Messages), Hash.length)
+             */
+
+            try {
+                // Use the ciphersuite's hashAlg for these calcs.
+                HashAlg hashAlg = cipherSuite.hashAlg;
+                KDF hkdf = KDF.getInstance(hashAlg.hkdfAlgorithm);
+
+                // First calculate the inner Derive-Secret(Secret, label, "")
+                MessageDigest md;
+                byte[] emptyHash;
+
+                // Create the "" digest...
+                try {
+                    md = MessageDigest.getInstance(hashAlg.toString());
+                    emptyHash = md.digest();
+                } catch (NoSuchAlgorithmException nsae) {
+                    throw new RuntimeException(
+                            "Hash algorithm " + cipherSuite.hashAlg.name +
+                                    " is not available", nsae);
+                }
+
+                // ...then the hkdfInfo...
+                byte[] hkdfInfo = SSLSecretDerivation.createHkdfInfo(
+                        hkdfInfoLabel, emptyHash, hashAlg.hashLength);
+
+                // ...then the "inner" HKDF-Expand-Label() to get the
+                // derivedSecret that is used as the Secret in the "outer"
+                // HKDF-Expand-Label().
+                SecretKey derivedSecret = hkdf.deriveKey("TlsKey",
+                        HKDFParameterSpec.expandOnly(exporterMasterSecret,
+                        hkdfInfo, hashAlg.hashLength));
+                try {
+                    // Now do the "outer" HKDF-Expand-Label.
+                    //     HKDF-Expand-Label(derivedSecret, "exporter",
+                    //         Hash(context_value), key_length)
+
+                    // If a context was supplied, use it, otherwise, use the
+                    // previous hashed value of ""...
+                    byte[] hash = ((context.length > 0) ?
+                            md.digest(context) : emptyHash);
+
+                    // ...now the hkdfInfo...
+                    hkdfInfo = SSLSecretDerivation.createHkdfInfo(
+                            ("tls13 exporter").getBytes(StandardCharsets.UTF_8),
+                            hash, length);
+
+                    // ...now the final expand.
+                    return (deriveKey ?
+                            hkdf.deriveKey("TlsExporterKeyingMaterial",
```

Comment:
Using a registered algorithm will make this method work, but I'm not sure if the resulting key can be used further. Do you know any typical use cases for the EKM?

Reply:
From a previous comment:

IIUC, the exported keying material can be used for any purpose or algorithm, so we really can't make an good educated guess what it might be.  They could be Keys (Ciphers), byte array/value challenges, or even just data that will be signed.  This is just doing a quick read of some of the IANA definitions which link to some of the known use cases:

    https://www.iana.org/assignments/tls-parameters/tls-parameters.xhtml#exporter-labels

Reply:
Personally, I would like to give user the chance to specify the algorithm themselves. A "TlsExporterKeyingMaterial" key will not be accepted by an AES cipher. If you are not ready for this, I'd rather only provide the `exportKeyingMaterialData` method now.

Reply:
How about adding a `String alg` parameter to `exportKeyingMaterialKey` like in the `KDF.deriveKey` API?

Reply:
As discussed with @seanjmullan / @wangweij , that is the direction I'll try.  It's not perfect, but a definite step in the right direction.  Thanks for the idea.

Reply:
Check the latest version, available after my build/test run.